### DISCUSSION TITLE
Build Python app with PyInstaller in Github Actions

### DISCUSSION THREAD

**Discussion by lrwagner** [community/community#25138] [Actions]
<p>Hi there,</p>
<p>Is it possible to run PyInstaller and push the created .exe File into my Repository?</p>
<p>I tried to implement a Github workflow, which runs PyInstaller on a .py File. </p>
<p>When I ran this script within my Github workflow I got the following error on the “build with pyinstaller” job :</p>
<p>Thanks in advance for your help :smiley:</p>
<pre><code>name: PyInstaller

on: [push]

jobs:
  build:

    runs-on: windows-latest

    steps:
    - uses: actions/checkout@v1
    - name: Set up Python 3.8
      uses: actions/setup-python@v1
      with:
        python-version: 3.8
    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install pyinstaller
    - name: build with pyinstaller
      run: |
        pyinstaller arbeitszeit.py

exec(code, run_globals)
  File "C:\hostedtoolcache\windows\Python\3.8.0\x64\Scripts\pyinstaller.exe\ __main__.py", line 7, in &lt;module&gt;
  File "c:\hostedtoolcache\windows\python\3.8.0\x64\lib\site-packages\PyInstaller\ __main__.py", line 111, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "c:\hostedtoolcache\windows\python\3.8.0\x64\lib\site-packages\PyInstaller\ __main__.py", line 63, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "c:\hostedtoolcache\windows\python\3.8.0\x64\lib\site-packages\PyInstaller\building\build_main.py", line 844, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "c:\hostedtoolcache\windows\python\3.8.0\x64\lib\site-packages\PyInstaller\building\build_main.py", line 791, in build
    exec(code, spec_namespace)
  File "D:\a\Test\Test\arbeitszeit.spec", line 18, in &lt;module&gt;
    pyz = PYZ(a.pure, a.zipped_data,
  File "c:\hostedtoolcache\windows\python\3.8.0\x64\lib\site-packages\PyInstaller\building\api.py", line 98, in __init__
    self. __postinit__ ()
  File "c:\hostedtoolcache\windows\python\3.8.0\x64\lib\site-packages\PyInstaller\building\datastruct.py", line 158, in __postinit__
    self.assemble()
  File "c:\hostedtoolcache\windows\python\3.8.0\x64\lib\site-packages\PyInstaller\building\api.py", line 128, in assemble
    self.code_dict = {
  File "c:\hostedtoolcache\windows\python\3.8.0\x64\lib\site-packages\PyInstaller\building\api.py", line 129, in &lt;dictcomp&gt;
    key: strip_paths_in_code(code)
  File "c:\hostedtoolcache\windows\python\3.8.0\x64\lib\site-packages\PyInstaller\building\utils.py", line 652, in strip_paths_in_code
    consts = tuple(
  File "c:\hostedtoolcache\windows\python\3.8.0\x64\lib\site-packages\PyInstaller\building\utils.py", line 653, in &lt;genexpr&gt;
    strip_paths_in_code(const_co, new_filename)
  File "c:\hostedtoolcache\windows\python\3.8.0\x64\lib\site-packages\PyInstaller\building\utils.py", line 660, in strip_paths_in_code
    return code_func(co.co_argcount, co.co_kwonlyargcount, co.co_nlocals, co.co_stacksize,
TypeError: an integer is required (got type bytes)
##[error]Process completed with exit code 1.
</code></pre>

**Comment by lrwagner** [comment, marked as answer]
<p>Seems like my issue was caused by <a href="https://github.com/pyinstaller/pyinstaller/issues/4265" rel="nofollow noopener">https://github.com/pyinstaller/pyinstaller/issues/4265</a></p>
<p>Using Python 3.7 instead of 3.8 fixed the problem for me</p>

**Comment by al-ra1** [comment]
<p><a class="mention" href="/lrwagner">@lrwagner</a>,</p>
<p>Thank you for posting the solution.  I am a relative newbie to programming in general.  I have been writing short programs in Python for last 6 months and genuinely enoying it.  Also I updated my Python version just last week from 3.7.3 to 3.8.0.  However, while trying to create my very first ever .exe file(s) from a .py/.pyw file(s) (across any version in any programming language), I ran into the issue of TypeError (integer required but got bytes).  I researched and quickly stumbled across your recommended solution.  However, would you PLEASE PLEASE PLEASE take the time to explain what to do after installing a separate copy of Python 3.7.5 (the one I just installed).  I don’t want to uninstall the 3.8.0 just yet.  What are the exact steps?  When I launch PowerShell (or Command Prompt) as Admin to create .exe, the 3.8 still shows up as my Python version.  Is there a command to manually switch to 3.7x when in PowerShell?  The 3.7.5 is working just fine if I use the Python IDLE / interpreter / iPython console in Spyder or PowerShell as a normal user.  Please explain the solution in simple English to a newbie.  Thank you much in advance!  Cheers</p>

**Comment by al-ra1** [comment]
<p><a class="mention" href="/lrwagner">@lrwagner</a> </p>
<p>I figured out how to set up 3.7x as the default system environment variable.  Thanks again for posting the solutiuon,</p>

**Comment by lrwagner** [comment]
<p>Hey <a class="mention" href="/al-ra1">@al-ra1</a> ,</p>
<p>I think what you are looking for is some sort of virtual environment for your projects, like venv, pipenv or conda.</p>
<p>It is considered best practice to use a new virtual env for each of your projects. During creation of your venv you can define which Python version you want to use. Let’s say you have Python 3.8 installed on your system, you can define any of 3.x versions for your virtual env. Additionally each project gets its own environment with all the third party libraries you want to install in it.</p>
<p>Consider reading through the following article: <a href="https://docs.python-guide.org/dev/virtualenvs/#basic-usage" rel="nofollow noopener">https://docs.python-guide.org/dev/virtualenvs/#basic-usage</a></p>

**Comment by al-ra1** [comment]
<p>Hi <a class="mention" href="/lrwagner">@lrwagner</a>,</p>
<p>Thank you for taking the time to answer my prior question (and sorry for this delayed response).  I mentioned shortly after sending my initial request, I was able to solve my issue by directly changing the environment variable itself.  I also considered creating a virtual environment at the time, but couldn’t decide if it was necessary.  I think next time on I’ll consider switching between the two environments right from the Command Prompt/PowerShell interpreter window.  Also thank you for attaching the link to Python documentation in your response.</p>
<p>Cheers</p>

**Comment by thomascury** [comment]
<p>Hi <a class="mention" href="/lrwagner">@lrwagner</a>,</p>
<p>I’m trying to build a automated packager for my app but I don’t know how to get the exe once it is generated by the Github action.</p>
<p>Ideally I would like to have it directly pushed back to the code repository.</p>
<p>Could you share how you did it ?</p>
<p>Regards.</p>
<p>Thomas</p>

**Comment by lrwagner** [comment]
<p>I think you can push back any File to your Repository from Github Actions using git in your action. Alternatively I would look at the possibility to create a new Release from within a Github Action and attaching the exe to it.</p>

**Comment by thomascury** [comment]
<p>Hey <a class="mention" href="/lrwagner">@lrwagner</a> , thank you for your answer.</p>
<p>I did try to push the exe back into the repository but git needs to authenticate to be able to commit &amp; push and I’m having a hard time authenticating in command-line only in the action workflow :confused:</p>
<p>Edit: I stumbled upon the poetry repository and their action workflows are a gold mine :smiley:</p>
<p>The short story is: they upload their packaged app as release assets using </p>
<pre><code>actions/upload-release-asset@v1
</code></pre>
